### PR TITLE
Update python image in task cluster

### LINKED ISSUE #629: Use python slim images where possible in .taskcluster.yml

### PR DESCRIPTION
**Pull request by Amotul-raheem**
fixes #629

### CODE REVIEW

File: .taskcluster.yml [hunk 1]
```diff
@@ -61,9 +61,9 @@ tasks:
           workerType: batch
           payload:
             maxRunTime: 3600
-            image: python:3.9
+            image: python:3.9-slim
             command:
-              - "/bin/bash"
+              - "/bin/sh"
```
Comment (suhaibmujahid) [inline .taskcluster.yml:66]: This seems causing issues. See: https://community-tc.services.mozilla.com/tasks/VKo0l4RKRPiPUo9EwvfFCQ/runs/0/logs/public/logs/live.log#L38

**Review by suhaibmujahid — COMMENTED** [review]
@Amotul-raheem feel free to covert back the PR when you think it is ready for review.

File: .taskcluster.yml (hunk 1 above)
Comment (Amotul-raheem) [reply .taskcluster.yml]: got it thank you. sorry for the trouble

Comment (suhaibmujahid) [reply .taskcluster.yml]: no trouble; thank you for the work!

**Comment by suhaibmujahid** [comment]
@Amotul-raheem I would advise covert the images one by one. Also, you do not need to convert all of them, only where possible. Installing the missed tools should not significantly impact the build time compared to using the full Python image.

Note: It could be easier for you to experiment locally (create a Docker file that reproduces the same steps).

**Comment by Amotul-raheem** [comment]
> @Amotul-raheem I would advise covert the images one by one. Also, you do not need to convert all of them, only where possible. Installing the missed tools should not significantly impact the build time compared to using the full Python image.
> 
> Note: It could be easier for you to experiment locally (create a Docker file that reproduces the same steps).

Yes, I did this exact same thing. I created a sample flask server running on docker and I **sh** into the running container using `docker exec it [container_id]  sh` to run test some of the commands to see what dependencies I need to install. 
I think everything is working fine now, but  bugbug tests status is [exception](https://community-tc.services.mozilla.com/tasks/JLwiYru7QtiOo_TapdNWSg) and I'm not sure of what to do. Maybe rerun the task again by making a dummy commit?

**Comment by Amotul-raheem** [comment]
@suhaibmujahid [all tasks](https://community-tc.services.mozilla.com/tasks/groups/ZxwCbM_WRdKGibLwIU5IeA) passed successfully.However, codecov dropped as expected. Please, how do I go about resolving this?

File: .taskcluster.yml [hunk 2]
```diff
@@ -153,8 +156,7 @@ tasks:
             - "-lcx"
             - "apt-get -qq update &&
               apt-get -qq install -y libhdf5-dev &&
-              git clone --quiet ${repository} &&
-              cd bugbug &&
```
Comment (marco-c) [inline .taskcluster.yml:157]: Nit: no need to move this to a single line

Comment (Amotul-raheem) [reply .taskcluster.yml:157]: @marco-c  and @suhaibmujahid Alright, will revert, thanks It's still failing code coverage. Can I add unnittest for [download_model.py ](https://github.com/mozilla/bugbug/blob/master/http_service/bugbug_http/download_models.py) file so as to bump the coverage? I'm thinking i should do it in another branch but it won't solve the problem as the branch lowers the code coverage.

File: .taskcluster.yml [hunk 3]
```diff
@@ -82,14 +84,20 @@ tasks:
...
             - "/bin/bash"
             - "-lcx"
             - "apt-get -qq update &&
               apt-get -qq install -y libhdf5-dev zstd &&
+              apt-get install -y git &&
+              apt-get install -y curl &&
+              apt-get install -y gcc &&
+              apt-get install -y python3-dev &&
+              apt-get install -y g++ &&
+              apt-get install -y mercurial &&
```
[the remark this answers is not in the document]
Comment (Amotul-raheem) [reply .taskcluster.yml]: done, thank you

File: .taskcluster.yml (hunk 2 above)
Comment (suhaibmujahid) [reply .taskcluster.yml]: > It's still failing code coverage. 

It is ok.

> Can I add unnittest for [download_model.py ](https://github.com/mozilla/bugbug/blob/master/http_service/bugbug_http/download_models.py) file so as to bump the coverage? I'm thinking i should do it in another branch but it won't solve the problem as the branch lowers the code coverage.

Yes, you could file an issue for that and open a PR to resolve it.

**Comment by suhaibmujahid** [comment]
@Amotul-raheem could you evaluate if using slim + apt update + apt install git is more or less efficient than just using the full image?

**Comment by Amotul-raheem** [comment]
> 

could you please explain what you mean by efficient here?

**Comment by Amotul-raheem** [comment]
> @Amotul-raheem could you evaluate if using slim + apt update + apt install git is more or less efficient than just using the full image?

could you please explain what you mean by efficient here?

**Comment by suhaibmujahid** [comment]
> could you please explain what you mean by efficient here?

I mean time to build (how fast)

**Comment by Amotul-raheem** [comment]
@suhaibmujahid I did the analysis and my findings are in the table below. 

| image  | build time | size |
| ------------- | ------------- | ------------- |
| python full image  | 1m 28.463s  | 915MB
| python-slim + apt update + apt install git | 19.566s  | 225MB 

I think python-slim is better as it has less build time and smaller image size

**Comment by suhaibmujahid** [comment]
Thank you, @Amotul-raheem! 

The time saving seems to be highly correlated with the image size. This indicates that the time saving is a reflection of the lower network traffic. I guess these results are based on a local test. I'm not sure if it will be the same on CI. If Taskcluster has any kind of caching or an on-prem image registry, traffic will have a smaller footprint on the overall result.

**Comment by Amotul-raheem** [comment]
> Thank you, @Amotul-raheem!
> 
> The time saving seems to be highly correlated with the image size. This indicates that the time saving is a reflection of the lower network traffic. I guess these results are based on a local test. I'm not sure if it will be the same on CI. If Taskcluster has any kind of caching or an on-prem image registry, traffic will have a smaller footprint on the overall result.

Interesting, thinking about it now it makes sense. 
Should we close the PR and the issue then?

**Comment by marco-c** [comment]
Agreed, timings on Taskcluster might be different. We can compare runs between master and the PR to see what the effect is.

**Comment by Amotul-raheem** [comment]
It made sense to add python slim package to only two tasks; the `bugbug lint test`and `bugbug tag version check`.

After triggering the task cluster using python slim and without python slim, the `bugbug lint test` executed with the same time,(3 minutes).

With this, i don't think using the python slim image is going to have any significant effect on the run time of these processes.

**Comment by Amotul-raheem** [comment]
@suhaibmujahid @marco-c , After these analysis, should the PR and issue be closed?

**Comment by suhaibmujahid** [comment]
Thank you for valuable contrbution here, @Amotul-raheem! I lean toward using the full Python images. A smaller image size could still be slightly benfitule in case we are pushing the built images to the registry. @marco-c wdyt?

**Comment by marco-c** [comment]
I agree, let's keep using the full Python image. Thank you anyway @Amotul-raheem, this was very useful investigation!

**Pull request closed without merging**